js/compiler.ts

```diff
@@ -0,0 +1,502 @@
+declare global {
+  interface Window {
+    define?: AmdDefine;
+  }
+}
+
+/**
+ * Abstraction of the APIs required from the `os` module so they can be
+ * easily substituted.
+ */
+export interface Os {
+  codeCache: typeof os.codeCache;
+  codeFetch: typeof os.codeFetch;
+  exit: typeof os.exit;
+}
+
+/**
+ * Abstraction of the APIs required from the `typescript` module so they can
+ * be easily substituted.
+ */
+export interface Ts {
+  createLanguageService: typeof ts.createLanguageService;
+  /* tslint:disable-next-line:max-line-length */
+  formatDiagnosticsWithColorAndContext: typeof ts.formatDiagnosticsWithColorAndContext;
+}
+
+/**
+ * A simple object structure for caching resolved modules and their contents.
+ *
+ * Named `FileModule` to clarify it is just a representation of meta data of
+ * the module, not the actual module instance.
+ */
+class FileModule {
+  scriptVersion = "";
+  readonly exports = {};
+
+  constructor(
+    public readonly fileName: string,
+    public readonly sourceCode = "",
+    public outputCode = ""
+  ) {
+    if (outputCode !== "") {
+      this.scriptVersion = "1";
+    }
+  }
+}
+
+/**
+ * The required minimal API to allow formatting of TypeScript compiler
+ * diagnostics.
+ */
+const formatDiagnosticsHost = {
+  getCurrentDirectory: () => ".",
+  getCanonicalFileName: (fileName: string) => fileName,
+  getNewLine: () => EOL
+};
+
+/**
+ * A singleton class that combines the TypeScript Language Service host API
+ * with Deno specific APIs to provide an interface for compiling and running
+ * TypeScript and JavaScript modules.
+ */
+export class DenoCompiler implements ts.LanguageServiceHost {
+  private readonly _fileModuleMap = new Map<string, FileModule>();
+  private readonly _fileNamesMap = new Map<string, Map<string, string>>();
+
+  // TODO ideally this are not static and can be influenced by command line
+  // arguments
+  private readonly _options: Readonly<ts.CompilerOptions> = {
+    allowJs: true,
+    module: ts.ModuleKind.AMD,
+    outDir: "$deno$",
+    inlineSourceMap: true,
+    inlineSources: true,
+    target: ts.ScriptTarget.ESNext
+  };
+
+  private _os: Os = os;
+  private _service: ts.LanguageService;
+  private _ts: Ts = ts;
+
+  /**
+   * Setup the global error handler for privileged bindings.
+   */
+  private _setupGlobalErrorHandler(): void {
+    // Uncaught exceptions are sent to window.onerror by the privileged binding.
+    window.onerror = (
+      message: string,
+      source: string,
+      lineno: number,
+      colno: number,
+      error: Error
+    ) => {
+      // TODO Currently there is a bug in v8_source_maps.ts that causes a
+      // segfault if it is used within window.onerror. To workaround we
+      // uninstall the Error.prepareStackTrace handler. Users will get unmapped
+      // stack traces on uncaught exceptions until this issue is fixed.
+      //Error.prepareStackTrace = null;
+      console.log(error.stack);
+      this._os.exit(1);
+    };
+  }
```

Comment:
Questionable if this should be part of the `DenoCompiler` class since it changes a global `window.onerror`... but if this is cleaner for you - ok.

Similarly for _setupSourceMaps below.

Reply:
Ok, I will think about it again...  It is almost seperate runtime setup stuff, which it stuck out, that didn't feel like it should go in a module named `compiler.ts` either.  Maybe we leave that stuff in runtime?  I started to feel a bit dirty about implicitly doing stuff anyways that wasn't related to the compiler.

Reply:
Upon looking at it, I moved the above to `global.ts`.  For `_setupSourceMaps` would be less clean though, as this code needs to be able to resolve modules based on the `fileName` to get the outputted code to unwind the stack traces.  It is certainly possible to do that "outside" of the class, but it would require grabbing the compiler instance, and any "special" modules would need to be maintained in the code.  So given that, I think it is best to keep it here for now until.  I will put a todo though about possibly rethinking that in the future.